ext/neural_snow/DataTools.jl

```diff
@@ -542,6 +633,1171 @@ function snotel_metadata(;
+        end
+        usedays = gooddays .& ((!).(flags))
+        if sum(usedays) < 30
+            continue
+        end
+        μ = mean(skipmissing(data[usedays]))
+        σ = std(skipmissing(data[usedays]))
+        zscores[months .== month] = (data[months .== month] .- μ) / σ
+        zflags[months .== month] = abs.(zscores[months .== month]) .> z
+    end
+    zflags[ismissing.(zflags)] .= true
+    return zscores, Vector{Bool}(zflags)
+end
+
+"""
+    yan_qc(input)
+
+Apply quality control for inconsistent water years as proposed by Yan et. al. (2018).
+
+**Note: requires column names of at least "date", "SWE", "precip" in the data
+
+# Arguments
+- `input::DataFrame`: the site data over which to apply quality control.
+"""
+function yan_qc(input::DataFrame)
+    data = deepcopy(input)
+    allowmissing!(data)
+    years = sort(unique(Dates.year.(data[!, :date])))
+    if Dates.month(data[end, :date]) >= 10
+        push!(years, years[end] + 1)
+    end
+    totdiff = 0.0
+    n_inconsistent = 0.0
+    numlost = 0
+    for year in years
+        oct1 = Date(string(year - 1) * "-10-01")
+        yearend = oct1 + Dates.Year(1)
+        allyear = oct1 .<= data[!, :date] .< yearend
+        if sum(allyear) == 0
+            continue
+        elseif sum(ismissing, data[allyear, :SWE]) == sum(allyear)
+            continue
+        end
+        (maxswe, maxidx) = findmax(skipmissing(data[allyear, :SWE]))
+        sweprec = data[allyear, :precip][maxidx]
+        if ismissing(sweprec)
+            data[allyear, [:SWE, :precip]] .= missing
+            continue
+        end
+        if maxswe > 1.05 * sweprec
+            data[allyear, [:SWE, :precip]] .= missing
+            totdiff += maxswe / sweprec
+            n_inconsistent += 1
+            numlost += sum(allyear)
+        end
+    end
+    avgdiff = totdiff / n_inconsistent
+    return data, avgdiff
+end
+
+"""
+    livneh_bc(input)
+
+Apply undercatch bias control for water years in the manner outlined by Livneh (2014),
+as is done by Yan et. al. (2019)
+
+**Note: requires column names of at least "date", "SWE", "precip" in the data, ordered sequentially.
+
+# Arguments
+- `input::DataFrame`: the site data over which to apply quality control.
+"""
+function livneh_bc(input::DataFrame)
+    data = deepcopy(input)
+    years = sort(unique(Dates.year.(data[!, :date])))
+    if Dates.month(data[end, :date]) >= 10
+        push!(years, years[end] + 1)
+    end
+
+    for year in years
+        oct1 = Date(string(year - 1) * "-10-01")
+        yearend = Date(string(year) * "-10-01")
+        precipdates = oct1 .<= data[!, :date] .< yearend
+        yeardata = data[precipdates, [:precip, :SWE]]
+        valid_idx = (!).(ismissing.(yeardata[!, :precip]))
+        valid_idx = valid_idx .& (!).(ismissing.(yeardata[!, :SWE]))
+        if sum(valid_idx) <= 2
+            continue
+        end
+        aprec = yeardata[valid_idx, :precip]
```

Comment:
what does "aprec" mean as a variable name?

Reply:
"aprec" = "accumulated precipitation" values. Can change to "acc_prec" if desired.

Reply:
Note: add high-level direction in comments to point to bcqc_hourly/daily, or "look at these for scraping", so that people don't have to read the whole file for one intended usage. "accum" better than "acc". Make variables coherent for code understanding instead of using paper values, use comment to map code vars to paper vars